File: structural/python/proxy/proxy.py
from subject import Subject


class Proxy(Subject):
    _real_subject: Subject = None

    def __init__(self, real_subject: Subject):
        self._real_subject = real_subject

    def request(self) -> None:
        if self.check_access():
            self._real_subject.request()
            self.log_access()

    def check_access(self) -> bool:
        print("Proxy: Checking access prior to firing a real request.")
        return True

    def log_access(self) -> None:
        print("Proxy: Logging the time of request.", end="")

File: structural/java/proxy/src/Proxy.java
public class Proxy implements Subject{
    private RealSubject realSubject;

    public Proxy(RealSubject realSubject){
        this.realSubject = realSubject;
    }

    @Override
    public void request() {
        if(this.checkAccess()){
            this.realSubject.request();
            this.logAccess();
        }
    }

    public boolean checkAccess(){
        System.out.println("Proxy: Checking access prior to firing a real request.");
        return true;
    }

    public void logAccess(){
        System.out.println("Proxy: Logging the time of request.");
    }
}
